### PR TITLE
boards: arm: rpi_pico: add openocd runner configuration

### PR DESCRIPTION
**Pull request by soburi**
Add openocd configuration.

This configuration allows selecting what debug adapter debugging rpi_pico
by a command-line argument.
It is mainly intended to support both the 'picoprobe' and 'raspberrypi-swd'
adapter described in "Getting started with Raspberry Pi Pico".
And any other SWD debug adapter might also be usable with this configuration.

### CODE REVIEW

**Review by yonsch — COMMENTED** [review]
Looking good overall, I have just one question:
Does this work with the openocd found in zephyr-sdk, or does it use a custom version of openocd? I couldn't find `target/rp2040.cfg` (though I did find `rp2040-core0.cfg`), and neither `interface/picoprobe`/`interface/raspberrypi-swd`.

**Comment by soburi** [comment]
Hi! @yonsch 
Thanks for reviewing.

> Looking good overall, I have just one question: Does this work with the openocd found in zephyr-sdk, or does it use a custom version of openocd? I couldn't find `target/rp2040.cfg` (though I did find `rp2040-core0.cfg`), and neither `interface/picoprobe`/`interface/raspberrypi-swd`.

It requires a custom version openocd.

I missed writing information about the distribution source of the custom version.
(https://github.com/raspberrypi/openocd)
I will add some information about that version.

There is an easy way to use this version of openocd is running the setup.sh script for setting up the pico development environment. I wrote it in the document.

**Comment by burumaj** [comment]
I tested this yesterday evening with the picoprobe. There are couple remarks I have regarding the documentation of this:

1) when using the picoprobe adding a `udev` rule might be necessary so a regular user can use it with openocd (or at least a user that is in de plugdev group in this case).
```
# Raspberry Pi Picoprobe
ATTRS{idVendor}=="2e8a", ATTRS{idProduct}=="0004", MODE="660", GROUP="plugdev", TAG+="uaccess"
```
2) When using an openocd not included with the zephyr sdk there is not rtos support for zephyr. The consequence of this is that it's possible to use openocd to debug an application, but only without thread awareness.

3) You mention the use of `picosetup.sh`, this script and is only supported on piOS, and probably works fine on debian based systems like ubuntu.

**Comment by soburi** [comment]
Hi, @burumaj 

> I tested this yesterday evening with the picoprobe. There are couple remarks I have regarding the documentation of this:

Thanks for testing it and comments.
I updated the document with apply your mentions.

**Review by yonsch — CHANGES_REQUESTED** [review]
Thanks, a few comments about the documentation.

File: boards/arm/rpi_pico/doc/index.rst [hunk 1]
```diff
@@ -83,17 +83,61 @@ Flashing
...
+How to connect `picoprobe` and `raspberrypi-swd` is described in `Getting Started with Raspberry Pi Pico`_.
+Any other SWD debug adapter maybe also work with this configuration.
+
+**RPI_PICO_DEBUG_ADAPTER** value remember into CMakeCache.txt.
+So you can omit the option in `west flash` and `west debug` execution,
+you need only the `west build` case.
+
+**RPI_PICO_DEBUG_ADAPTER** is used in an argument to openocd as `"source [find interface/${RPI_PICO_DEBUG_ADAPTER}.cfg]"`.
+Thus, **RPI_PICO_DEBUG_ADAPTER** needs to assign from the definition file name of debugging adapter.
+
```
[the remark this answers is not in the document]
Comment (soburi) [reply boards/arm/rpi_pico/doc/index.rst:131]: Fix it

File: boards/arm/rpi_pico/doc/index.rst [hunk 2]
```diff
@@ -83,17 +83,61 @@ Flashing
...
+   ATTRS{idVendor}=="2e8a", ATTRS{idProduct}=="0004", MODE="660", GROUP="plugdev", TAG+="uaccess"
+
+This example is valid for the case that the user joins to `plugdev` groups.
+
 The Raspberry Pi Pico has an SWD interface that can be used to program
 and debug the on board RP2040. This interface can be utilized by openocd.
-However, to use it with the RP2040, a custom fork of openocd is needed.
+However, to use it with the RP2040, `fork of OpenOCD supporting RP2040`_ is needed.
 This fork can be found here: https://github.com/raspberrypi/openocd
```
[the remark this answers is not in the document]
Comment (soburi) [reply boards/arm/rpi_pico/doc/index.rst]: I removed this line.

File: boards/arm/rpi_pico/doc/index.rst [hunk 3]
```diff
@@ -112,8 +156,22 @@ Debugging
 =========
 
 The SWD interface can also be used to debug the board. To achieve this,
-install openocd as described for flashing the board. Also, install gdb-multiarch.
-Then run the following command:
+install openocd as described for flashing the board.
```
[the remark this answers is not in the document]
Comment (soburi) [reply boards/arm/rpi_pico/doc/index.rst:152]: I wrote on "Flashing" section previous push, but "Debugging" section is more suitable.
Move it to here.

**Comment by MeisterBob** [comment]
I would suggest adding jlink support to board.cmake as well

```
board_runner_args(jlink "--device=RP2040_M0_0")
include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
```

**Comment by soburi** [comment]
Hi, @MeisterBob 

Thanks for comment.

> I would suggest adding jlink support to board.cmake as well
> 
> ```
> board_runner_args(jlink "--device=RP2040_M0_0")
> include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
> ```

It seems good.
And this code is your contribution.
So, I think it is better to create your PR.
(In another view, it keeps simple PR.)

Please consider it.

**Comment by soburi** [comment]
@MeisterBob ,
I split the topic about J-Link support to new issue.
https://github.com/zephyrproject-rtos/zephyr/issues/48194

(Simply reason, I can't verify the code within this PR becasuse I don't have J-Link debugger...)

**Review by yonsch — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**